src/vmm/src/vstate.rs

```diff
@@ -283,6 +296,100 @@ pub struct Vcpu {
 }
 
 impl Vcpu {
+    thread_local!(static TLS_VCPU_PTR: VcpuCell = Cell::new(None));
+
+    /// Associates `self` with the current thread.
+    ///
+    /// Is a prerequisite to succesfully run `run_on_thread_local()` on the current thread.
+    /// Will fail if there is already a `Vcpu` present in this thread's TLS.
+    fn init_thread_local_data(&mut self) -> Result<()> {
+        Self::TLS_VCPU_PTR.with(|cell: &VcpuCell| {
+            if cell.get().is_some() {
+                return Err(Error::VcpuTlsInit);
+            }
+            cell.set(Some(self as *const Vcpu));
+            Ok(())
+        })
+    }
+
+    /// Deassociates `self` from the current thread.
+    ///
+    /// Should be called if the current `self` had called `init_thread_local_data()` and
+    /// now needs to move to a different thread.
+    ///
+    /// Fails if `self` was not previously associated with the current thread.
+    fn reset_thread_local_data(&mut self) -> Result<()> {
+        // Best-effort to clean up TLS. If the `Vcpu` was moved to another thread
+        // _before_ running this, then there is nothing we can do.
+        let mut tls_contains_self = false;
+        Self::TLS_VCPU_PTR.with(|cell: &VcpuCell| {
+            if let Some(vcpu_ptr) = cell.get() {
+                if vcpu_ptr == self as *const Vcpu {
+                    tls_contains_self = true;
+                }
+            }
+        });
+
+        if tls_contains_self {
+            Self::TLS_VCPU_PTR.with(|cell: &VcpuCell| cell.take());
+            Ok(())
+        } else {
+            Err(Error::VcpuTlsNotPresent)
+        }
+    }
```

Comment:
This function can be shorter as below:
```rust
    fn reset_thread_local_data(&mut self) -> Result<()> {
        // Best-effort to clean up TLS. If the `Vcpu` was moved to another thread
        // _before_ running this, then there is nothing we can do.
        Self::TLS_VCPU_PTR.with(|cell: &VcpuCell| {
            if let Some(vcpu_ptr) = cell.get() {
                if vcpu_ptr == self as *const Vcpu {
                    Self::TLS_VCPU_PTR.with(|cell: &VcpuCell| cell.take());
                    return Ok(());
                }
            }
            Err(Error::VcpuTlsNotPresent)
        })
    }
```

Reply:
Niiiice! I hadn't tried that as I thought it is double borrowing, but it's actually not.
Cool! Done.